pyHalo/realization_extensions.py

```diff
@@ -328,4 +377,102 @@ def add_primordial_black_holes(self, pbh_mass_fraction, kwargs_pbh_mass_function
+
+        rmin_kpc,rmax_kpc = args['rmin'] * to_kpc, args['rmax'] * to_kpc #args in kpc
+        area_ring = np.pi*(rmax_kpc**2-rmin_kpc**2) # volume of ring
+        n_flucs=int(area_ring/fluc_area) # number of fluctuations in ring
+    
+    if shape=='ellipse': # fluctuations in a elliptical slice (for visualization purposes)
+        
+        amin_kpc,bmin_kpc,amax_kpc,bmax_kpc=args['amin'] * to_kpc, args['bmin'] * to_kpc, args['amax'] * to_kpc, args['bmax'] * to_kpc #args in kpc
+        area_ellipse=np.pi*(amax_kpc*bmax_kpc - amin_kpc*bmin_kpc) # volume of ellipse
+        n_flucs=int(area_ellipse/fluc_area) # number of fluctuations in ellipse
+
+    if shape=='aperture': # fluctuations around lensing images (for computation)
+
+        n_images=len(args['x_images']) #number of lensed images
+        r_kpc = args['aperture'] * D_d * arcsec * 1e3 #aperture in kpc
+        area_aperture = np.pi*r_kpc**2 # aperture area
+        n_flucs_expected = area_aperture/fluc_area #number of expected fluctuations per aperture
+        n_flucs = np.random.poisson(n_flucs_expected,n_images) #draw number of fluctuations from poisson distribution for each image
+        n_flucs[np.where(n_flucs == 0)] = 1 #avoid zero fluctuations
+
+    return n_flucs
+
+def _get_fluctuation_halos(realization,de_Broglie_wavelength,rho_bar,shape,n_flucs,args):
+    """
+    This function creates 'n_flucs' Gaussian fluctuations and places them according to 'shape'.
+
+    :param realization: the realization to which to add the fluctuations
+    :param de_Broglie_wavelength: de Broglie wavelength of the DM particle in kpc
+    :param rho_bar: typical convergence of a fluctuation at its peak
+    :param shape: keyword argument for fluctuation geometry, see 'add_ULDM_fluctuations'
+    :param n_flucs: Number of fluctuations to make
+    :param args: properties of the given shape, see 'add_ULDM_fluctuations'
+    """
+
+    D_d = realization.lens_cosmo.cosmo.D_A_z(realization._zlens)
+    arcsec = realization.lens_cosmo.cosmo.arcsec
+    r_dB_angle = de_Broglie_wavelength / D_d / arcsec / 1e3 #de Broglie wavelength in arcsec
+
+    mass = RealizationDefaults().host_m200
+    c = realization.lens_cosmo.NFW_concentration(mass, realization._zlens)
+    _,_,r200 = realization.lens_cosmo.NFW_params_physical(mass,c,realization._zlens)
+    CLT_factor = np.sqrt(2*de_Broglie_wavelength/3/r200)
+
+    if shape!='aperture':
+
+        sigs = np.abs(np.random.normal(r_dB_angle,r_dB_angle/2,n_flucs)) #random widths
+        amps = CLT_factor*np.random.normal(0,np.sqrt(2)*rho_bar,n_flucs)*(2*np.pi*sigs**2) #random amplitudes
+
+    if shape=='ring':
+
+        angles = np.random.uniform(0,2*np.pi,n_flucs)  # random angles
+        radii = args['rmin'] + np.sqrt(np.random.uniform(0,1,n_flucs))*(args['rmax']-args['rmin']) #random radii
+        xs = radii*np.cos(angles) #random x positions
+        ys = radii*np.sin(angles) #random y positions
+
+    if shape=='ellipse':
```

Comment:
can you add some documentation to describe how you determine the orientation of the ellipse?

Reply:
In what sense? How the paramter args['angle'] determines the inclination of the ellipse? Or the equations I use to convert from cartesian to polar?

Reply:
ah nevermind, I missed that you have to pass in 'angle' to args. maybe you can add that to the documentation in the add_ULDM_fluctuations function

Reply:
Ok sure